Fix #589 updated node dependencies and devDependencies

badrmodoukh opened the pull request:
I updated the dependencies and devDependencies in the package.json file. I used [npm-check](https://www.npmjs.com/package/npm-check) as suggested by @humphd to help with the update. These are the packages that needed to be updated.
<img width="876" alt="screen shot 2017-02-24 at 12 03 18 pm" src="https://cloud.githubusercontent.com/assets/12388934/23314909/d9d9f064-fa91-11e6-96ed-c0974393fe0f.png">

The only package I didn't update is autoprefixer-core. There was an update to version 6.0.1 but when I tried updating it I got an error when testing.
<img width="645" alt="screen shot 2017-02-24 at 12 36 37 pm" src="https://cloud.githubusercontent.com/assets/12388934/23315083/b413d538-fa92-11e6-9ef9-056838bf97ed.png">

I also ignored the updates for the dependencies/devDependencies that Adobe uses. As stated by @humphd these will get updated when we merge their upstream commits later.

Pomax commented:
I'm a bit worried about this PR because it highlights the fact that we don't have any testing in place right now to detect whether or not any of these updates actually (subtly or not) break Thimble. We can do manual testing, but ideally any manual test is only run once: to make sure it can be automated.

I'm okay with merging this in but we should have a fast follow up that starts to introduce all the tests that were done manually for this PR in an automated fashion so we can repeat them.

/cc @gideonthomas @humphd

badrmodoukh commented:
Is this comment directed to me or someone else?
If so, I tested this by installing Bramble again with these commands:
- npm install
- npm run build
- npm start
Which made it available at http://localhost:8000/src/hosted.html 
I also ran Thimble by doing vagrant up and it worked fine

I'm not sure what other tests I need to perform

Pomax commented:
@badrmodoukh it's a general comment to everyone looking at this PR, as well as gideon and humph specifically. We don't have an npm test command, so you technically didn't test anything, all you were able to do (and we're on the hook for that, that's not a failing on your part) is determine that you were able to start brackets without terminal errors. That is not enough to be able to say that things actually work, there could be any number of bugs that crept in with these updates that we'll only run into during very specific use cases that require much more extensive QA testing beyond starting up brackets =(

I don't think in this case there are any specific test you can do in addition to this, but some discussion is needed on whether we consider this PR safe enough to land when it updates quite a few packages to beyond the major the versions they were.

humphd commented:
See also https://github.com/mozilla/thimble.mozilla.org/issues/1606 for the issue of testing.  We probably need to do it here too, just for Bramble.  But that should be a new bug.

humphd commented:
@badrmodoukh what's up with the `postcss:dist` failure in your grunt run?  Looks like something broke with your update.

badrmodoukh commented:
@humphd sorry about that I did not pay attention to that failure. This `postcss:dist` failure happen when I tried installing Bramble having updated `autoprefixer-core` to version 6.0.1 and `postcss` to ^0.8.0

Installing Bramble with `autoprefixer-core` being version 5.1.8 and `postcss` being ^0.8.0 I get this message

<img width="515" alt="screen shot 2017-02-26 at 2 46 22 pm" src="https://cloud.githubusercontent.com/assets/12388934/23343101/14c1b48e-fc34-11e6-9934-394f4e77e810.png">

Changing the version of `postcss` to the previous version 0.3.0 runs without any failures and shows this

<img width="394" alt="screen shot 2017-02-26 at 3 03 05 pm" src="https://cloud.githubusercontent.com/assets/12388934/23343153/d7cc7284-fc34-11e6-8845-7561c6a8a6a2.png">

I should update this pull request to have `postcss` be version 0.3.0

humphd commented:
Looks like the issue is `autoprefixer-core` itself being deprecated: https://github.com/ai/autoprefixer-core.

It looks to me like you need to not only update the modules, but switch over to `autoprefixer` and change the code in our Gruntfile.js for the `postcss` task to be more like what's here:

https://github.com/nDmitry/grunt-postcss#usage

Often updating module dependencies means rewriting code to work with the new APIs they use.

badrmodoukh commented:
@humphd I updated `autoprefixer` and `postcss` and changed over to autoprefixer. I also changed the code for the `postcss` task in the Gruntfile.js 

I reinstalled Bramble and got this message for `postcss:dist`

<img width="521" alt="screen shot 2017-02-26 at 5 07 41 pm" src="https://cloud.githubusercontent.com/assets/12388934/23344209/375c19e6-fc46-11e6-9335-3dd17e89b5c8.png">

Everything seems to work fine on my end now

humphd commented on package.json:
```diff
@@ -1,82 +1,82 @@
 {
-    "name": "bramble",
```
You've converted this file to 2-space indent from 4-space indent.  Can you change it back please?  It's important, since that's what Adobe does upstream, and if we want to merge with them in future, we need to not have git think we changed every line.

humphd commented on package.json:
```diff
@@ -1,82 +1,82 @@
+    "grunt-git": "^0.3.4",
+    "grunt-jasmine-node": "0.1.0",
+    "grunt-lib-phantomjs": "0.3.0",
+    "grunt-npm": "git://github.com/sedge/grunt-npm.git#branchcheck",
```
@gideonthomas do you know if we can dump this hack yet?

gideonthomas replied:
Yeah I'm pretty sure. I actually think we can even remove it entirely because the grunt task that uses it, we don't really use anymore.

humphd commented:
@badrmodoukh can I confirm with you that you've left the deps that Adobe uses upstream alone?

https://github.com/adobe/brackets/blob/master/package.json

@gideonthomas do you agree we should leave shared ones where Adobe has them, or update?

Pomax commented:
modern autoprefixing with postcss involves using `postcss` in combination with `browserslist`, which automatically hooks into postcss to make prefixing happen.

humphd commented:
@gideonthomas are there any others you can see that we should dump?

badrmodoukh commented:
@humphd Yes, I left the deps that Adobe uses upstream alone

I also fixed the indent to 4-space indent, sorry about that

humphd commented on package.json:
```diff
@@ -14,59 +14,59 @@
+        "autoprefixer": "^6.0.0",
+        "aws-sdk": "^2.19.0",
         "chokidar": "1.6.0",
-        "lodash": "4.15.0",
```
Why are these being removed?  I don't want the shared Adobe deps removed, just left at the version they are already at.  We need all these.

NOTE: you not only need to see that your grunt tasks run after you change things, but you should also try running the app.  You'll find that removing stuff will break things.

humphd replied:
Oh, it looks like they just got moved.  nvm

gideonthomas commented:
I'll take a look at what we can remove in a bit.

badrmodoukh commented:
@gideonthomas I just merged your PR

gideonthomas commented:
This looks good to me! I'll leave the final sign-off to you @humphd

gideonthomas commented:
@badrmodoukh thanks for doing this, really nice work!
